Add strategy in the auth function signatures (#296)
kuzzleio/documentation: src/core/1/guide/guides/essentials/real-time/index.md

---
code: false
type: page
title: Real-time Engine
order: 600
---

# Real-time Engine

Kuzzle includes his own real-time engine for sending notifications to clients connected through the API.

Real-time capabilities requires the use of a persistent communication protocol such as WebSocket or MQTT.

> Kuzzle offers 2 different ways to make real time:

Reviewer: ```suggestion
Kuzzle offers 2 different ways of doing real-time:
```

 - volatile Pub/Sub system
 - real-time database notifications

 ::: info

> You can bypass notification triggering by using [bulk:write](/core/1/api/api-reference/bulk-controller/write) or [bulk:mWrite](/core/1/api/api-reference/bulk-controller/m-write) actions.

Reviewer: ```suggestion
You can bypass notifications from being triggered by using actions from the [bulk controller](/core/1/api/api-reference/bulk-controller).
```

:::

### wscat

This guide provides examples that use the Kuzzle API directly through a command line WebSocket client: [wscat](https://github.com/websockets/wscat).

To install it you can type the following command: `npm install -g wscat`.

The sample requests are to be sent directly to wscat after connecting to your Kuzzle server with the command `wscat --connect localhost:7512`.

It is of course possible to send the payloads provided with a different WebSocket client than wscat.

## Pub/Sub

> Kuzzle's real-time engine allows you to do Pub/Sub in dedicated communication channels called room.

Reviewer: ```suggestion
Kuzzle's real-time engine allows you to do Pub/Sub in dedicated communication channels called rooms.
```

The process is as follows:
 - a customer subscribes to a particular room,
 - a second customer posts a message in this room,
 - the customer subscribing to the room receives a notification.

![kuzzle-pub-sub](./pub-sub.png)

> Subscription to a room is done via the [realtime:subscribe](/core/1/api/api-reference/controller-realtime/subscribe) method, it takes 3 parameters that will be used to define a specific room:

Reviewer: ```suggestion
Subscription to a room is done via the [realtime:subscribe](/core/1/api/api-reference/controller-realtime/subscribe) method. It takes 3 parameters, used to describe a specific room:
```

 - name of an index,
 - name of a collection,
 - subscription filters contained in the `body` of the request.

::: info

> For using Kuzzle in Pub/Sub mode only, the index and collection do not need to physically exist in the database (e.g. be created in Kuzzle via the <a href="/core/1/api/api-reference/controller-index/create">index:create</a> and <a href="/core/1/api/api-reference/controller-collection/create">collection:create</a> methods of the API).

Reviewer 2: ```suggestion
In order to use Kuzzle in Pub/Sub mode only, the index and collection do not need to physically exist in the database (e.g. be created in Kuzzle via the <a href="/core/1/api/api-reference/controller-index/create">index:create</a> and <a href="/core/1/api/api-reference/controller-collection/create">collection:create</a> methods of the API).
```

> In order to use Kuzzle in Pub/Sub mode only, the index and collection do not need to physically exist in the database (e.g. be created in Kuzzle via the <a href="/core/1/api/api-reference/controller-index/create">index:create</a> and <a href="/core/1/api/api-reference/controller-collection/create">collection:create</a> methods of the API).

Reviewer: ```suggestion
In order to use Kuzzle in Pub/Sub mode only, the index and collection do not need to physically exist in the database (e.g. created in Kuzzle via the <a href="/core/1/api/api-reference/controller-index/create">index:create</a> and <a href="/core/1/api/api-reference/controller-collection/create">collection:create</a> methods of the API).
```

```

Customers subscribing to this channel will receive the following notification:

<details><summary>Click to expand</summary>
<pre>
{
  "status": 200,
  "requestId": "644bc890-9c14-4a8f-afcc-afef444fd6f7",
  "timestamp": 1558690506519,
  "volatile": null,
  "index": "nyc-open-data",
  "collection": "yellow-taxi",
  "controller": "realtime",  // Controller who trigger the notification
  "action": "publish",       // Action who trigger the notification
  "protocol": "websocket",
  "scope": "in",
  "result": {
    "_source": {
      // Content of the published document
      "name": "Manwë",
      "licence": "B",
      "car": "berline",
      "position": {
        "lat": 43.6073913,
        "lon": 3.9109057
      },
      "_kuzzle_info": {
        "author": "-1",
        "createdAt": 1558690506527
      }
    },
    "_id": null
  },
  "type": "document", // Notification is of type document
  "room": "54ffaa49fe470879bed9b0697468bb21-89c22fbf000567a2ed2e7886ed0c51e3"
}
</pre>
</details>

More information about the [Document Notification format](/core/1/api/essentials/notifications/#documents-changes-messages)

::: warning

> Messages published with the <a href="/core/1/api/api-reference/controller-realtime/publish">realtime:publish</a> method are not persistent in the database.

Reviewer: ```suggestion
Messages published with the <a href="/core/1/api/api-reference/controller-realtime/publish">realtime:publish</a> method are not persisted in the database.
```

:::

## Database notifications

Kuzzle's real-time engine allows you to subscribe to notifications corresponding to changes in the database.

> Subscription to a database changes is done via the [realtime:subscribe](/core/1/api/api-reference/controller-realtime/subscribe) method, it takes 3 parameters:

Reviewer: ```suggestion
Subscription to a database changes is done via the [realtime:subscribe](/core/1/api/api-reference/controller-realtime/subscribe) method, taking 3 parameters:
```

 - name of the index,
 - name of the collection you want to watch,
 - subscription filters contained in the `body` of the request.

::: info
The specified index and collection must exist in the database to receive database notifications.
:::

> When changes occur on this collection (eg: document creation, modification or deletion), Kuzzle will send notifications to notify the different subscribers.

Reviewer: ```suggestion
When changes occur on this collection (eg: document creation, modification or deletion), Kuzzle will send notifications to the corresponding subscribers.
```

These filters allow the customer to tell Kuzzle exactly which documents they are interested in and only receive notifications about them.

::: info
These filters are specified only on the client side and do not require server-side implementation.
They are sent in the body of the request [realtime:subscribe](/core/1/api/api-reference/controller-realtime/subscribe)
:::

A filter is composed of [term](/core/1/koncorde/essentials/terms) that can be composed with [operands](/core/1/koncorde/essentials/operands).

For example if I want to receive only drivers with the `B` license:
```json
{
  "controller": "realtime",
  "action": "subscribe",
  "index": "nyc-open-data",
  "collection": "yellow-taxi",
  "body": {
    "equals": { "licence": "B" }
  }
}
```

It is also possible to combine [terms](/core/1/koncorde/essentials/terms) between them with [operands](/core/1/koncorde/essentials/operands) to refine my filter:

```json
{
  "controller": "realtime",
  "action": "subscribe",
  "index": "nyc-open-data",
  "collection": "yellow-taxi",
  "body": {
    "and": [
      { "equals": { "licence": "B" } },
      { "equals": { "car": "berline" } }
    ]
  }
}
```

> Each subscription filter define a scope. All documents in the collection can be either inside or outside this scope.

Reviewer: ```suggestion
Each subscription filter defines a scope. All documents in the collection can be either inside or outside this scope.
```

> Once a customer has subscribed to notifications with filters, he will receive notifications each time a document enters or exits the scope defined by the filters.

Reviewer: ([explanation](https://en.wikipedia.org/wiki/Singular_they))
```suggestion
Once a customer has subscribed to notifications with filters, they will receive notifications each time a document enters or exits the scope defined by the filters.
```

 - `out`: receive only notifications about documents exiting the scope
 - `all`: (default) receive everything

For example, to be informed of taxis arriving at central park:
```json
{
  "controller": "realtime",
  "action": "subscribe",
  "index": "nyc-open-data",
  "collection": "yellow-taxi",
  "body": {
    "geoBoundingBox": {
      "position": {
        "topLeft": { "lat": 40.759507, "lon": -73.985384 },
        "bottomRight": { "lat": 40.758372, "lon": -73.984591 }
      }
    }
  },
  "scope": "in" // Only documents entering the scope
}
```

### users

The [users](/core/1/api/api-reference/controller-realtime/subscribe/#arguments) option allows you to receive additional notifications when another client joins or leaves the same room.

This parameter can take 4 values:
 - `in`: only receive notifications when a customer joins the room
 - `out`: only receive notifications when a customer leaves the room
 - `all`: receive everything
 - `none`: (default) receive nothing

```json
 {
   "controller": "realtime",
   "action": "subscribe",
   "index": "nyc-open-data",
   "collection": "yellow-taxi",
   "body": {
   },

>    "user": "all" // Receive notification when user enter or exit the room

Reviewer: ```suggestion
   "user": "all" // Receive notification when a user enters or exits the room
```
